### ISSUE TITLE
 ERROR DryRunExecutor :  TypeError: Cannot set properties of undefined (setting 'currentTestId') 

### ISSUE THREAD

**Issue by bodote** [stryker-mutator/stryker-js#3780]
**Summary**

wenn stryker-js is run with config listed below on  this project:
```bash
git clone git@github.com:bodote/book-monkey.git
git checkout v0.2
```
I get this error during DryRunExecution on all tests:
```
...
08:41:57 (37435) INFO DryRunExecutor Starting initial test run (karma test runner with "perTest" coverage analysis). This may take a while.
08:41:57 (37435) DEBUG DryRunExecutor Using timeout of 300000 ms.
08:41:57 (37435) DEBUG TimeoutDecorator Starting timeout timer (300000 ms) for a test run
08:41:58 (37435) TRACE ChildProcessProxy Chrome Headless 105.0.5195.125 (Mac OS 10.15.7): Executed 0 of 134 SUCCESS (0 secs / 0 secs)

08:41:58 (37435) TRACE ChildProcessProxy Chrome Headless 105.0.5195.125 (Mac OS 10.15.7) FormMessagesComponent should create an array of errormessage that fit to a "title" control  FAILED
	TypeError: Cannot set properties of undefined (setting 'currentTestId')
	    at Object.specStarted (node_modules/@stryker-mutator/karma-runner/dist/src/karma-plugins/test-hooks-middleware-21f23d35-a4c9-4b01-aeff-da9c99c3ffc0.js:5:44)
	    at <Jasmine>
	    at ZoneQueueRunner.jasmine.QueueRunner.ZoneQueueRunner.execute (node_modules/zone.js/fesm2015/zone-testing.js:692:42)
	    at <Jasmine>
Chrome Headless 105.0.5195.125 (Mac OS 10.15.7): Executed 1 of 134 (1 FAILED) (skipped 25) (0 secs / 0.018 secs)

...
``` 

Whereas if I replace `"src/app/app.component.ts",` in the stryker.conf.json by  ` "src/app/shared/*.ts"`
It works without errors.
And althought of cause the `ng test` for src/app/app.component.ts or for the hole project runs also without any hickups

**Stryker config**

```json
{
  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
  "_comment": "This config was generated using 'stryker init'. Please see the guide for more information: https://stryker-mutator.io/docs/stryker-js/guides/angular",
  "mutate": [
    "src/app/app.component.ts",
    "!src/**/*.spec.ts",
    "!src/test.ts",
    "!src/environments/*.ts",
    "!src/**/*.module.ts",
    "!main.ts",
    "!cypress/**"
  ],
  "testRunner": "karma",
  "karma": {
    "configFile": "karma.conf.js",
    "projectType": "angular-cli",
    "config": {
      "browsers": [
        "ChromeHeadless"
      ]
    }
  },
  "reporters": [
    "progress",
    "clear-text",
    "html"
  ],
  "htmlReporter": { "fileName": "reports/mutation/mutation.app.html" },
  "concurrency": 2,
  "concurrency_comment": "Recommended to use about half of your available cores when running stryker with angular",
  "coverageAnalysis": "perTest",
  "timeoutFactor": 40,
  "checkers": ["typescript"],
  "fileLogLevel": "info",
  "tsconfigFile": "tsconfig.stryker.json",
  "tempDirName": "stryker-tmp",
  "inPlace": true
}

```

**Test runner config**

```JavaScript
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      jasmine: {
        // you can add configuration options for Jasmine here
        // the possible options are listed at https://jasmine.github.io/api/edge/Configuration.html
        // for example, you can disable the random execution with `random: false`
        // or set a specific seed with `seed: 4321`
      },
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    jasmineHtmlReporter: {
      suppressAll: true // removes the duplicated traces
    },
    coverageReporter: {
      dir: require('path').join(__dirname, './coverage/book-monkey'),
      subdir: '.',
      reporters: [
        { type: 'html' },
        { type: 'text-summary' }
      ]
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_DEBUG,
    autoWatch: true,
    browsers: ['ChromeHeadless'],
    singleRun: false,
    restartOnFileChange: true
  });
};
```

```
├── @stryker-mutator/core@6.2.2
├── @stryker-mutator/karma-runner@6.2.2
├── @stryker-mutator/typescript-checker@6.2.2
```

```
├── karma-chrome-launcher@3.1.1
├── karma-coverage@2.2.0
├── karma-jasmine-html-reporter@2.0.0
├── karma-jasmine@5.1.0
├── karma@6.4.1

```

**Test runner environment**

```shell
ng test
```

karma.conf.json see above
tsconfig.stryker.json:
```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./out-tsc/app",
    "types": []
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.d.ts"
  ],
  "exclude": [
    "src/**/*.spec.ts"
  ]
}
``` 

**Your Environment**

| software         | version(s)
| ---------------- | -------
| node             |v18.10.0
| npm              |v8.19.2
| Operating System |MacOS 12.6

**Add stryker.log**
see attached
[stryker.log](https://github.com/stryker-mutator/stryker-js/files/9738585/stryker.log)

**Comment by bodote** [comment]
found the reason: 
app.component.ts has only a constructor, but no other methods. Nevertheless there is logic in the html-template, so the 
app.component.spec.ts indeed contains 2 tests.

As soon as I add any method to app.component.ts, stryker does no longer throw this error

**Comment by nicojs** [comment]
Thanks for opening this issue! Love to have a reproduction repo 😍. I think I know what's going on here.

This was the smoking gun:

>  Instrumented 1 source file(s) with 0 mutant(s)

So there are 0 mutants. That means the `__stryker__` global variable also doesn't get introduced in the source code. The Stryker karma plugin expects the `__stryker__` global variable to be there, because it injects this piece of code:

```js
jasmine.getEnv().addReporter({
  specStarted: function (spec) {
    window.__stryker__.currentTestId = spec.id;
  }
});
```

I think an easy fix is to add this line:

```diff
+ window.__stryker__ = window.__stryker__ || {};
  jasmine.getEnv().addReporter({
    specStarted: function (spec) {
      window.__stryker__.currentTestId = spec.id;
    }
  });
```

As a workaround: As long as there is at least 1 file with 1 mutant, Stryker should work as expected.

**Comment by nicojs** [comment]
For anyone interested in picking this up: https://github.com/stryker-mutator/stryker-js/blob/339a3f248d7a78a66b3b5f6fe170d1926d4d88bc/packages/karma-runner/src/karma-plugins/test-hooks-middleware.ts#L86-L112

**Comment by bodote** [comment]
I try...

**Comment by bodote** [comment]
unfortunately the "Integration tests core" show error even before I change something in the code (I forked this project , added `git remote add upstream git@github.com:stryker-mutator/stryker-js.git`
and run `git fetch upstream && git merge upstream/master && git push && npm i` to be up to date.)
```
56 passing (2m)
/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/node_modules/mocha/lib/reporters/base.js:379
  1 failing
/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/node_modules/mocha/lib/reporters/base.js:392
  1) ChildProcessProxy
       should throw an OutOfMemoryError if the process went out-of-memory:

      AssertionError: expected promise to be rejected with 'OutOfMemoryError' but it was rejected with 'Error: Child process [pid 94460] exit…'
      + expected - actual

      -Error: Child process [pid 94460] exited unexpectedly with exit code null (SIGABRT). Last part of stdout and stderr was:
      -	59: 0x1050dc004 v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	60: 0x104ea8b94 node::InternalCallbackScope::Close() [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	61: 0x104ea8fd4 node::InternalMakeCallback(node::Environment*, v8::Local<v8::Object>, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	62: 0x104ec3630 node::AsyncWrap::MakeCallback(v8::Local<v8::Function>, int, v8::Local<v8::Value>*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	63: 0x1050252bc node::StreamBase::CallJSOnreadMethod(long, v8::Local<v8::ArrayBuffer>, unsigned long, node::StreamBase::StreamBaseJSChecks) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	64: 0x10502553c node::EmitToJSStreamListener::OnStreamRead(long, uv_buf_t const&) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	65: 0x10502b66c node::LibuvStreamWrap::OnUvRead(long, uv_buf_t const*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	66: 0x10502b35c node::LibuvStreamWrap::ReadStart()::$_1::__invoke(uv_stream_s*, long, uv_buf_t const*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	67: 0x1075d7004 uv__stream_io [/opt/homebrew/Cellar/libuv/1.44.2/lib/libuv.1.dylib]
      -	68: 0x1075de1e0 uv__io_poll [/opt/homebrew/Cellar/libuv/1.44.2/lib/libuv.1.dylib]
      -	69: 0x1075ce7d0 uv_run [/opt/homebrew/Cellar/libuv/1.44.2/lib/libuv.1.dylib]
      -	70: 0x104ea9878 node::SpinEventLoop(node::Environment*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	71: 0x104fb37dc node::NodeMainInstance::Run(int*, node::Environment*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	72: 0x104fb3430 node::NodeMainInstance::Run() [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	73: 0x104f3da10 node::LoadSnapshotDataAndRun(node::SnapshotData const**, node::InitializationResult*) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	74: 0x104f3db00 node::Start(int, char**) [/opt/homebrew/Cellar/node/18.10.0/bin/node]
      -	75: 0x1074e508c 
      -	
      +OutOfMemoryError
```. 
I have MacOS 12.6. with Apple M1 with 16Gbyte RAM

**Comment by bodote** [comment]
...and the integration tests for the karma-runner also give me an error: 
`mocha --timeout 18000 --exit "dist/test/integration/**/*.js"`:
```
  34 passing (25s)
  1 failing

  1) Infinite loop
       should report a timeout eventually and be able to recover from it:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'survived'
- 'timeout'
      + expected - actual

      -survived
      +timeout
      
      at Module.expectTimeout (file:///Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/test-helpers/src/assertions.ts:34:10)
      at Context.<anonymous> (file:///Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/karma-runner/test/integration/timeout-on-infinite-loop.it.spec.ts:35:18)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)

```

**Comment by bodote** [comment]
but apart from that : 
the fix proposed from @nicojs  seems to work.
so I added a line to the unit test to cover this change and  this is the 
[pull request 3787](https://github.com/stryker-mutator/stryker-js/pull/3787)

**Comment by nicojs** [comment]
Closed with #3787
